Frameworks/CoreGraphics/DWriteWrapper.mm

```diff
@@ -832,35 +398,29 @@ static HRESULT __DWriteUpdateUserCreatedFontCollection(CFArrayRef datas, CFArray
     // Still want to update font collection with whatever fonts succeeded, so return ret at end
     HRESULT ret = func(loader.Get(), datas, errors);
 
-    // DWrite won't create a new font collection unless we provide a new key each time
-    // So every time we modify the font collection increment the key
-    static size_t collectionKey = 0;
-    std::lock_guard<std::mutex> guard(_userCreatedFontCollectionMutex);
-    RETURN_IF_FAILED(
-        dwriteFactory->CreateCustomFontCollection(loader.Get(), &(++collectionKey), sizeof(collectionKey), &_userCreatedFontCollection));
-
-    // Update s_userFontPropertiesMap with new values
-    s_userFontPropertiesMap = __CreatePropertiesMapForFontCollection(_userCreatedFontCollection.Get());
+    // Update the user font collection
+    std::lock_guard<std::recursive_mutex> guard(s_userFontCollectionLock);
```

Comment:
read-write access to the whole font collection helper. if we locked in UpdateCollection(), we could still call ->CopyFontFamilyNames() from it, for instance. (we _could_ alleviate that by overriding CopyFontFamilyNames() and the other read functions to also lock, but I don't find that to be preferable to the current pattern of a lock around the whole data structure)

---
In reply to: [90936157](https://github.com/Microsoft/WinObjC/pull/1504#discussion_r90936157) [](ancestors = 90936157)